### PR TITLE
First version of element

### PR DESCRIPTION
**Pull request by bblaszkow06**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,4 +20,68 @@ def deps do
...
+
+  alias Membrane.Element.{File, FLACParser, IBMSpeechToText}
+
+  def handle_init(_) do
+    children = [
+      src: %File.Source{location: "sample.flac"},
+      parser: FLACParser,
+      sink: %IBMSpeechToText.Sink{
+        region: :frankfurt,
+        api_key: "ADD_YOUR_API_HERE"
```
Comment (mat-hek) [inline README.md]: *API_KEY

File: README.md [hunk 2]
```diff
@@ -14,4 +20,68 @@ def deps do
 end
 ```
 
-The docs can be found at [HexDocs](https://hexdocs.pm/membrane_element_ibm_speech_to_text).
+## Usage
+
+The input stream for this element should be parsed, so most of the time it should be
+placed in pipeline right after FLACParser
```
Comment (mat-hek) [inline README.md]: link to parser

File: lib/ibm_speech_to_text/sink.ex [hunk 3]
```diff
@@ -1,23 +1,110 @@
...
+        :input,
+        %Buffer{payload: payload, metadata: %FLAC.FrameMetadata{} = meta},
+        _ctx,
+        %{connection: conn, sample_rate: sample_rate, start_time: start_time} = state
+      ) do
+    Client.send_data(conn, payload)
+
+    next_sample_num = meta.starting_sample_number + meta.samples
+    next_frame_time = start_time + trunc(next_sample_num * Time.seconds(1) / sample_rate)
+    demand_time = (next_frame_time - Time.os_time()) |> max(0) |> Time.to_milliseconds()
```
Comment (mat-hek) [inline lib/ibm_speech_to_text/sink.ex]: such stuff should be added to caps

File: lib/ibm_speech_to_text/sink.ex [hunk 4]
```diff
@@ -1,23 +1,110 @@
...
   @impl true
   def handle_prepared_to_playing(_context, state) do
-    {{:ok, demand: :input}, state}
+    with {:ok, pid} <- Client.start_link(state.region, state.api_key) do
+      {{:ok, demand: :input}, %{state | connection: pid}}
+    end
+  end
+
+  @impl true
+  def handle_caps(:input, %FLAC{} = caps, _, %{connection: conn} = state) do
```
Comment (mat-hek) [inline lib/ibm_speech_to_text/sink.ex]: _ctx

File: lib/ibm_speech_to_text/sink.ex [hunk 5]
```diff
@@ -1,23 +1,110 @@
...
+    Client.send_message(conn, %Message.Start{content_type: :flac, interim_results: true})
+    start_time = Time.os_time()
+    state = %{state | start_time: start_time, sample_rate: caps.sample_rate}
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_event(:input, %EndOfStream{}, _, %{connection: conn} = state) do
```
Comment (mat-hek) [inline lib/ibm_speech_to_text/sink.ex]: _ctx

File: lib/ibm_speech_to_text/sink.ex [hunk 6]
```diff
@@ -1,23 +1,110 @@
...
+      response.results
+      |> Enum.filter(fn result -> result.final end)
+      |> Enum.map(fn result ->
+        alternative = result.alternatives |> hd()
+        alternative.transcript
+      end)
+
+    case transcipts do
+      [] -> {:ok, state}
+      _ -> {{:ok, notify: transcipts}, state}
```
Comment (mat-hek) [inline lib/ibm_speech_to_text/sink.ex]: send to a pid instead

File: lib/ibm_speech_to_text/sink.ex [hunk 7]
```diff
@@ -1,23 +1,110 @@
...
+  def handle_write(:input, %Buffer{payload: payload}, _ctx, %{connection: conn} = state) do
+    Client.send_data(conn, payload)
+    {{:ok, demand: :input}, state}
+  end
+
+  @impl true
+  def handle_other(%Response{} = response, _ctx, state) do
+    transcipts =
+      response.results
+      |> Enum.filter(fn result -> result.final end)
```
Comment (mat-hek) [inline lib/ibm_speech_to_text/sink.ex]: do this filter in the lib

File: README.md [hunk 8]
```diff
@@ -14,4 +20,68 @@ def deps do
...
+
+    spec = %Membrane.Pipeline.Spec{
+      children: children,
+      links: links
+    }
+
+    {{:ok, spec}, %{}}
+  end
+
+  def handle_notification(notification, elem, state) do
```
Comment (mat-hek) [inline README.md]: print transcripts only

File: lib/ibm_speech_to_text/sink.ex [hunk 9]
```diff
@@ -1,23 +1,110 @@
-defmodule Membrane.Element.IbmSpeechToText.Sink do
+defmodule Membrane.Element.IBMSpeechToText.Sink do
```
Comment (mat-hek) [inline lib/ibm_speech_to_text/sink.ex]: add some docs

File: lib/ibm_speech_to_text/sink.ex [hunk 10]
```diff
@@ -1,23 +1,110 @@
...
+        %Buffer{payload: payload, metadata: %FLAC.FrameMetadata{} = meta},
+        _ctx,
+        %{connection: conn, sample_rate: sample_rate, start_time: start_time} = state
+      ) do
+    Client.send_data(conn, payload)
+
+    next_sample_num = meta.starting_sample_number + meta.samples
+    next_frame_time = start_time + trunc(next_sample_num * Time.seconds(1) / sample_rate)
+    demand_time = (next_frame_time - Time.os_time()) |> max(0) |> Time.to_milliseconds()
+    Process.send_after(self(), :demand_frame, demand_time)
```
Comment (mat-hek) [inline lib/ibm_speech_to_text/sink.ex]: maybe add option for non-live processing?

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/ibm_speech_to_text/sink.ex (hunk 7 above)
Comment (bblaszkow06) [reply lib/ibm_speech_to_text/sink.ex]: Nope, as someone might need the interim results (not final)

File: lib/ibm_speech_to_text/sink.ex (hunk 10 above)
Comment (bblaszkow06) [reply lib/ibm_speech_to_text/sink.ex]: Maybe not in this version

File: lib/ibm_speech_to_text/sink.ex (hunk 3 above)
Comment (bblaszkow06) [reply lib/ibm_speech_to_text/sink.ex]: As discussed, this deserves a separate package for common audio stuff (or maybe should be part of the core?). Anyway, I'll leave it as-is for now

**Review by mat-hek — APPROVED** [review]

**Pull request merged**